streamalert/shared/normalize.py

```diff
@@ -13,22 +13,224 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 """
+from collections import defaultdict
 import logging
+import itertools
 
 from streamalert.shared.config import TopLevelConfigKeys
+from streamalert.shared.exceptions import ConfigError
 from streamalert.shared.logger import get_logger
 
 
 LOGGER = get_logger(__name__)
 LOGGER_DEBUG_ENABLED = LOGGER.isEnabledFor(logging.DEBUG)
 
+class NormalizedType:
+    """The class encapsulates normalization information for each normalized type"""
+
+    VALID_KEYS = {'fields', 'function'}
+    CONST_STR = 'str'
+    CONST_DICT = 'dict'
+
+    def __init__(self, log_type, normalized_type, params):
+        """Init NormalizatedType
+        Args:
+            log_type (str): log type name, e.g. osquery:differential
+            normalized_type (str): Normalized type name defined in conf/, e.g. 'sourceAddress',
+                'destination_ip' may be normalized to 'ip_address'.
+            params (list): a list of str or dict contains normalization configuration.
+                When it is read from conf/normalized_type.json, the params is a list of str, e.g.
+                    ['ipv4', 'local_ip']
+                When it is read from conf/schemas/*.json, the params should be a list of dictionary
+                having following format, otherwise it will raise ConfigError.
+                    [
+                        {
+                            'fields': ['source', 'sourceIPAddress'],
```

Comment:
This is not how I intended it to work. In this documentation you're implying that each normalizer can map to multiple fields. The way I designed it is each normalizer maps to **_exactly one_** field. The `field` array is a JSON path.

Reply:
```
[
  {
    'fields': ['path', 'to', 'the', 'field'],
    'function': 'same_function'
  },
  {
    'fields': ['other', 'path', 'same', 'function'],
    'function': 'same_function'
  },
]
```

Reply:
Yeah, misunderstood the configuration part in the design doc. The new change will be up soon.